### DISCUSSION TITLE
Running the code

### DISCUSSION THREAD

**Discussion by juribk** [boris-kz/CogAlg#55] [Q&A]
For a better understanding of your idea, I want to run the code from the repository.
But for some reason I run into problems. 
For example, when I run line_patterns.py, I get an error.
**line_patterns.py**
_PPm_ = **search**(P_)_ - not found (there is a similar function in line_PPs_draft.py)
**line_PPs_draft.py** - class_cluster does not contain comp_param
_from class_cluster import ClusterStructure, NoneType, **comp_param**, comp_param_complex_
I did not find **comp_param(param, _param, param_name)_** in the repository.
There is a similar **comp_param(param, _param, param_name, ave)_**, but it contains more parameters
There are other problems in the code.
What should I do to make the code from the repository work?

**Comment by kwcckw** [comment]
Hi, 

You may just remove comp_param and comp_param_complex, since ClusterStructure file for 1D may not updated to the one in 2D folder:

        import numpy as np
        from line_patterns import *
        from class_cluster import ClusterStructure, NoneType
        
 Then, you should be able to run the file, i just tried it.

**Reply by boris-kz** [reply]
Sorry, there is no need for ClassStructure in 1D alg, it should be imported from 2D alg.
I keep forgetting to do that.

**Comment by juribk** [comment]
I did as you advised. But it did not help.
I run **line_1D_alg\line_patterns.py**

**line_patterns.py** _-call->_ **line_PPs_draft.py\search(P_)** _-call->_ **line_PPs_draft.py\comp_P(P, _P, neg_M, neg_L)** _-call->_ **comp_param(param, _param, param_name)**

And again I cannot find **comp_param(param, _param, param_name)**

I think I am doing something wrong.
Where am I going wrong?

**Comment by kwcckw** [comment]
Did you remove the comp_param and comp_param_complex from the import process?

It should be like this:

![image](https://user-images.githubusercontent.com/52521979/120891203-dd1c7f80-c639-11eb-9952-4d4720051cbc.png)

**Reply by juribk** [reply]
Yes i did it

**Reply by juribk** [reply]
![image](https://user-images.githubusercontent.com/19360711/120891290-6d45cf00-c610-11eb-9c51-978631794d8e.png)

**Reply by juribk** [reply]
![image](https://user-images.githubusercontent.com/19360711/120891314-90707e80-c610-11eb-925c-fbb9b546289c.png)

**Reply by kwcckw** [reply]
Right now we are use Pycharm to debug the code, are you using Pycharm? If you are using some other IDE, probably there are some settings that we are not aware of.

**Reply by juribk** [reply]
I am not using Pycharm. I'll install it now.
But this problem should not depend on the IDE.
Anyway, I'll install Pycharm now and try

**Reply by kwcckw** [reply]
Right, most of the time the problem should be independent on the IDE, but could you check again whether you are using the latest version of code?

**Reply by juribk** [reply]
I cloned the repository this morning

**Reply by juribk** [reply]
I have installed Pycharm.
The problems are the same

**Reply by juribk** [reply]
![image](https://user-images.githubusercontent.com/19360711/120892557-a33a8180-c617-11eb-984d-84e033e328e3.png)
![image](https://user-images.githubusercontent.com/19360711/120892579-c107e680-c617-11eb-9574-f77b9982a9ce.png)
![image](https://user-images.githubusercontent.com/19360711/120892609-e0067880-c617-11eb-9a45-f0837f1e5991.png)
![image](https://user-images.githubusercontent.com/19360711/120892640-ff050a80-c617-11eb-96c4-76de52993658.png)

**Reply by kwcckw** [reply]
Okay, could you try import class_cluster from other location instead?

    from frame_2D_alg.class_cluster import ClusterStructure, NoneType, comp_param,comp_param_complex

Then, make sure these are checked:

![image](https://user-images.githubusercontent.com/52521979/120892853-c0d11080-c642-11eb-8f51-0823898d4b60.png)

Also do take note you need to clone the whole repo, including the 2D algo folder as well.

**Reply by juribk** [reply]
It looks like these are two different functions
**def comp_param(param, _param, param_name, ave)** - three parameters
**d, m = comp_param(param, _param, param_name)** - two parameters

**Reply by juribk** [reply]
I think it will be faster if you send me a project by mail. 
Is it possible?

**Reply by boris-kz** [reply]
Yes, sorry, I messed it up yesterday. But you don't need it to run line_patterns, just exclude line_PPs_draft and related.

> It looks like these are two different functions
> **def comp_param(param, _param, param_name, ave)** - three parameters
> **d, m = comp_param(param, _param, param_name)** - two parameters

**Reply by juribk** [reply]
![image](https://user-images.githubusercontent.com/19360711/120893367-9cae0900-c61b-11eb-801a-d78f5c6548e9.png)
Yes, It worked!
Now I will see what it does)))

**Reply by boris-kz** [reply]
Great, sorry for the trouble!

**Reply by boris-kz** [reply]
I got your pdf @juribk, appreciate you interest!  
But it seems that getting into this project will require a large investment of time, from both of us.  
Which won't happen unless you are interested in the idea. And you mentioned that you don't understand it enough, because it is deeply theoretical in nature. So, help me understand please, what is it that made you interested, enough to justify such investment?
Best regards.

**Reply by juribk** [reply]
Thanks for giving me the chance!
My strong point is testing, optimization and implementation. Since I do not fully understand the deep theory of the idea, I can offer myself as a tester.
The first thing that comes to my mind is to put the CogAlg method into practice.
In my main job, I develop equipment for oil rigs. One of the tasks is to transmit data through the hydrochannel. Signal decoding is required here. I have already implemented the basic part in ML.
I can suggest trying out CogAlg algorithms for decoding tasks. If I understand correctly this is Line_1d_alg.
For this task I can use all my resources.

**Reply by boris-kz** [reply]
Thanks. Unfortunately, we can't put CogAlg into practice or even properly test it, because it's far from complete.
Even 1D alg: line_patterns is only the first level, line_PPs (2nd level) is a work in progress, and only after that's finished we can start working on a level-recursive version. 

Which may do something interesting, or we may have to translate it into 2D or 3D level-recursive alg first. Ours is a 4D universe, lower-D patterns may not be very meaningful on their own. So, it's a long way from practice and optimization. And then decoding probably won't be a good initial use case, as it is inherently symbolic.

**Reply by juribk** [reply]
I got it.
Then it would be nice if you give me an assignment.
This will allow me to get the necessary starting knowledge about your project.

**Reply by boris-kz** [reply]
Ok great. So, the first thing is to understand line_patterns, it's the most basic code.
I know you already run it, but did you see the output patterns? 

It's best to do in Pycharm: set init_y = 500 and put a breakpoint in line 80 because you only need to process one row in the image, that's why it's 1D. Then step to the next row and Pycharm variables tab in debug window will show P_: a row of patterns formed by processing previous row 500 of pixels. Then you can unfold P_ into individual Ps, and each P into multiple levels of variables, sequentially.
BTW, Pycharm has powerful breakpoints, just right-click on one and you can type any sort of code in the window, as a condition.

We are not drawing the output because it's largely symbolic, you need to see variables. For that, Variables tab in Pycharm debug window is far better.

**Comment by boris-kz** [comment]
> Hi, Boris.
I am Marat, who was introduced by Mr. Chris. I have read CogAlg several times and tried to understand.
When line_1D_alg.py was executed, it showed me execution time, not anything else. I am not sure how it works.
Furthermore, I hope you let me know what I do to run frame_2D_alg. There are many main.
Of course, I understand what CogAlg does. and it is very interesting to me.
Thanks.

Great, appreciate your interest Marat!
So, line_patterns is best understood by running it in Pycharm debug window and checking Variables tub, 
please see my reply above.
Same for frame_blobs, but please wait till I push some fixes tonight.

**Comment by alex-pitertsev** [comment]
Hi, Boris and Chris. Sorry for the delay. I would like to propose my initial code review.
[Alex's review.pdf](https://github.com/boris-kz/CogAlg/files/6718773/Alex.s.review.pdf)

**Reply by boris-kz** [reply]
Thanks Alex! I use plain Python in cross_comp on purpose, it's simpler to understand for most people and introduces notation that is used across the project. Your numpy version is shorter, but I don't think it's clearer. Speed makes no difference here, we only need to process one line at a time.

But we do use numpy in 2D alg, please check comp_pixel in frame_blobs and intra_comp.
Re output visualization, Yuri is working on it, should be posting shortly.

Thanks again, do you think you understood the rest of line_patterns?

**Reply by alex-pitertsev** [reply]
Thanks for the answer, Boris. I guess I have understood what the rest of the line_patterns code does. I will have a look at frame_blobs and intra_comp files. The main problem for me during the working on math description of the project is to formulate how to check the results of processing by the cognitive algorithm, whether are right or wrong.

**Reply by boris-kz** [reply]
Well, the processing is cross-comp and clustering. Right or wrong is a matter of consistency of operations with definitions. It operations are correct, then so must be the results, there is no way to get "correct" results otherwise.

So, I think the only place where math descriptions may make sense is general definitions of comparison, in Readme, "Atomic comparison: quantifying match and miss between two variables" section.
Maybe also latter when we will have discontinuous clustering, that will look like graphs.

**Reply by alex-pitertsev** [reply]
I guess I have understood the idea about the cross-comp and clustering, thank you. 
I am sorry, but I would like to express my disagreement with such a position about the correctness of results. I believe that results of data processing are correct when they permit to demonstrate the solution of any kind of task (classification, prediction, detection, or something other.). This solution could be far from perfect, slow, not very accurate so on, but it should exist. I am not trying to be critical, I would like to make a contribution in a form of an example of such a demonstration. To show the principle vitality of cognitive algorithm. You have already a quite solid code base and I guess this step should be important enough for the development of the project.

**Reply by boris-kz** [reply]
Appreciate your opinion, but our code base is far from solid, we change things all the time. And we do so based on theoretical consistency rather than experimentation. Interpretation of experimentation in DL is fundamentally subjective, such as with using human-labeled images in ImageNet. And it is fundamentally very coarse, nothing happens until algorithm is trained on countless images with 50000 cycles of backprop. 

None of this can be done when core algorithm is <10% complete, we haven't even started coding our version of feedback, which would replace ANN training.

**Comment by alex-pitertsev** [comment]
Hello, I would like to show my questions and propositions for 1D. I tried to add a little data visualization also.
This is pdf created from the jupyter notebook with a slightly modified 1D version. I could share a live version through google colab, for example. I have used # +++ comments to highlight new strings
[line_1D_alg_modified.pdf](https://github.com/boris-kz/CogAlg/files/6795928/line_1D_alg_modified.pdf)

**Comment by Maksimkhokhovk** [comment]
Hello boris, I'm Maksim.
I tried to go through your algorithm and I think it's good way for me to execute your main code on each Dimension.
So I am successful on 1D and understand about your code.
But I tried frame_blobs.py on 2D and got error.
It's same bellow.
"ImportError: cannot import name 'Cdm' from 'class_cluster'"
I want to be solved this problem.
Thanks.

**Reply by boris-kz** [reply]
Sorry Maksim, Cdm was recently changed to Cdert. But you don't need it to run frame_blobs, I will remove it.
